File: web/src/api/knowledge.ts
// web/src/api/knowledge.ts
import request from './request'

// ========== 类型定义 ==========

export interface NovelKnowledge {
  id: number
  novel_id: number
  category: string
  title: string
  content: string
  tags: string
  chapter_ref: string
  priority: number
  status: string
  sort_order: number
  resolved: boolean
  created_at: string
  updated_at: string
}

// 知识类别选项
export const knowledgeCategories = [
  { value: 'character', label: '人物档案', icon: '👤' },
  { value: 'worldview', label: '世界观设定', icon: '🌍' },
  { value: 'plotline', label: '剧情线索', icon: '📖' },
  { value: 'foreshadow', label: '伏笔追踪', icon: '🔮' },
  { value: 'style', label: '文风规范', icon: '✒️' },
  { value: 'custom', label: '自定义', icon: '📌' },
] as const

// ========== API ==========

export const knowledgeApi = {
  // 获取小说的知识条目列表
  list: (novelId: number, params?: { category?: string; status?: string }) =>
    request.get(`/novels/${novelId}/knowledge`, { params }),

  // 创建知识条目
  create: (novelId: number, data: {
    category: string
    title: string
    content: string
    tags?: string
    chapter_ref?: string
    priority?: number
  }) => request.post(`/novels/${novelId}/knowledge`, data),

  // 获取知识条目详情
  get: (kid: number) => request.get(`/knowledge/${kid}`),

  // 更新知识条目
  update: (kid: number, data: Partial<{
    category: string
    title: string
    content: string
    tags: string
    chapter_ref: string
    priority: number
    status: string
  }>) => request.put(`/knowledge/${kid}`, data),

  // 删除知识条目
  delete: (kid: number) => request.delete(`/knowledge/${kid}`),

  // 确认待审核条目
  confirm: (kid: number) => request.post(`/knowledge/${kid}/confirm`),

  // 批量确认
  batchConfirm: (novelId: number) =>
    request.post(`/novels/${novelId}/knowledge/batch-confirm`),

  // 搜索知识条目
  search: (novelId: number, keyword: string) =>
    request.get(`/novels/${novelId}/knowledge/search`, { params: { keyword } }),

  // AI 提取知识
  extract: (novelId: number, data: { chapter_id: number; model_name?: string }) =>
    request.post(`/novels/${novelId}/knowledge/extract`, data),

  // 解析 AI 提取结果
  parseExtract: (novelId: number, data: { chapter_id: number; task_id: number }) =>
    request.post(`/novels/${novelId}/knowledge/parse-extract`, data),
}

File: server/internal/model/knowledge.go
// server/internal/model/knowledge.go
package model

import "time"

// NovelKnowledge 小说知识条目表，支持结构化知识管理
type NovelKnowledge struct {
	ID         uint      `gorm:"primaryKey" json:"id"`
	NovelID    uint      `gorm:"not null;index:idx_novel_category" json:"novel_id"`
	Category   string    `gorm:"size:30;not null;index:idx_novel_category" json:"category"` // character/worldview/plotline/foreshadow/style/custom
	Title      string    `gorm:"size:200;not null" json:"title"`
	Content    string    `gorm:"type:text;not null" json:"content"`
	Tags       string    `gorm:"size:500" json:"tags"`        // 逗号分隔的标签，用于关键词匹配
	ChapterRef string    `gorm:"size:500" json:"chapter_ref"` // 关联章节 ID 列表，逗号分隔
	Priority   int       `gorm:"default:0;index" json:"priority"`
	Status     string    `gorm:"size:20;not null;default:confirmed" json:"status"` // confirmed/pending
	SortOrder  int       `gorm:"default:0" json:"sort_order"`                     // 情节线排序
	Resolved   bool      `gorm:"default:false" json:"resolved"`                   // 伏笔回收标记
	CreatedAt  time.Time `json:"created_at"`
	UpdatedAt  time.Time `json:"updated_at"`
}

// 知识条目类别枚举
const (
	KnowledgeCategoryCharacter  = "character"  // 人物档案
	KnowledgeCategoryWorldview  = "worldview"  // 世界观设定
	KnowledgeCategoryPlotline   = "plotline"   // 主线/支线剧情
	KnowledgeCategoryForeshadow = "foreshadow" // 伏笔追踪
	KnowledgeCategoryStyle      = "style"      // 文风规范
	KnowledgeCategoryCustom     = "custom"     // 自定义
)

// ValidKnowledgeCategories 合法的知识类别白名单
var ValidKnowledgeCategories = map[string]bool{
	KnowledgeCategoryCharacter:  true,
	KnowledgeCategoryWorldview:  true,
	KnowledgeCategoryPlotline:   true,
	KnowledgeCategoryForeshadow: true,
	KnowledgeCategoryStyle:      true,
	KnowledgeCategoryCustom:     true,
}

// 知识条目状态枚举
const (
	KnowledgeStatusConfirmed = "confirmed" // 用户确认
	KnowledgeStatusPending   = "pending"   // AI 提取待审核
)

// ValidKnowledgeStatuses 合法的知识状态白名单
var ValidKnowledgeStatuses = map[string]bool{
	KnowledgeStatusConfirmed: true,
	KnowledgeStatusPending:   true,
}

// KnowledgeCategoryLabel 类别中文标签（用于 Prompt 注入时的分组标题）
var KnowledgeCategoryLabel = map[string]string{
	KnowledgeCategoryCharacter:  "人物档案",
	KnowledgeCategoryWorldview:  "世界观设定",
	KnowledgeCategoryPlotline:   "剧情线索",
	KnowledgeCategoryForeshadow: "伏笔追踪",
	KnowledgeCategoryStyle:      "文风规范",
	KnowledgeCategoryCustom:     "自定义设定",
}
